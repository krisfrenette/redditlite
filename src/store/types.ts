import {ActionNames} from 'store/action-names';

export type SubredditState = {
    name: string;
    posts: RedditPosts;
    before: string;
    after: string;
    count: number;
    currentPaging: {
        direction: PagingDirection;
        count: number;
    };
    fetchingData: boolean;
};

export interface IRootState {
    subreddit: SubredditState;
}

export type RedditPost = {
    id: string;
    author: string;
    title: string;
    url: string;
    created: number;
    commentUrl: string;
    thumbnail?: Thumbnail;
};

export type Thumbnail = {
    url: string;
    width: number;
    height: number;
};

export type PagingDirection = {
    direction: 'before' | 'after';
    value: string;
};

export type RedditPosts = RedditPost[];

export type SubredditActions = {
    FETCH_NEW_POSTS_REQUEST: {
        type: typeof ActionNames.FETCH_NEW_POSTS_REQUEST;
    };
    FETCH_NEW_POSTS_SUCCESS: {
        type: typeof ActionNames.FETCH_NEW_POSTS_SUCCESS;
        payload: {
            before: string;
            after: string;
            posts: RedditPosts;
            currentPaging: {
                direction: PagingDirection;
                count: number;
            };
        };
    };
    RESET_STATE: {
        type: typeof ActionNames.RESET_STATE;
    };
    FETCH_NEW_POSTS_FAILED: {
        type: typeof ActionNames.FETCH_NEW_POSTS_FAILED;
    };
};

export type RootAction = SubredditActions[keyof SubredditActions];
