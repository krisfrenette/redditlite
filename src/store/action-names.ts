export enum ActionNames {
    FETCH_NEW_POSTS_REQUEST = 'subreddit/fetch/new/request',
    FETCH_NEW_POSTS_SUCCESS = 'subreddit/fetch/new/success',
    FETCH_NEW_POSTS_FAILED = 'subreddit/fetch/new/failed',
    RESET_STATE = 'subreddit/reset'
}
