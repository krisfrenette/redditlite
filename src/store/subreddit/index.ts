import {SubredditState, RootAction} from 'store/types';
import {ActionNames} from 'store/action-names';

const getDefaultState = (): SubredditState => {
    return {
        name: null,
        posts: [],
        before: null,
        after: null,
        count: 0,
        currentPaging: null,
        fetchingData: true
    };
};

export default function reducer(
    state: SubredditState = getDefaultState(),
    action: RootAction
): SubredditState {
    switch (action.type) {
        case ActionNames.FETCH_NEW_POSTS_REQUEST:
            return {
                ...state,
                fetchingData: true
            };
        case ActionNames.FETCH_NEW_POSTS_SUCCESS:
            return {
                ...state,
                fetchingData: false,
                ...action.payload
            };
        case ActionNames.FETCH_NEW_POSTS_FAILED:
            return {
                ...state,
                fetchingData: false
            };
        case ActionNames.RESET_STATE:
            return getDefaultState();

        default:
            return state;
    }
}
