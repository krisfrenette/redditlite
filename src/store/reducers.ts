import {combineReducers} from 'redux';
import subreddit from 'store/subreddit';

export default combineReducers({
    subreddit
});
