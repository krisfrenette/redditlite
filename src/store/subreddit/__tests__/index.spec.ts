import reducer from 'store/subreddit';
import {ActionNames} from 'store/action-names';

describe('reducer', () => {
    it('sets a default state', () => {
        expect(reducer(undefined, {type: 'foobar'} as any)).toMatchSnapshot();
    });

    it('handles the fetch posts request action', () => {
        expect(
            reducer({} as any, {
                type: ActionNames.FETCH_NEW_POSTS_REQUEST
            })
        ).toMatchSnapshot();
    });

    it('handles the fetch posts success action', () => {
        expect(
            reducer({} as any, {
                type: ActionNames.FETCH_NEW_POSTS_SUCCESS,
                payload: {
                    before: null,
                    after: null,
                    posts: [{id: 'mockId'}],
                    currentPaging: {
                        direction: null,
                        count: 0
                    }
                } as any
            })
        ).toMatchSnapshot();
    });

    it('handles the fetch posts failed action', () => {
        expect(
            reducer({} as any, {
                type: ActionNames.FETCH_NEW_POSTS_FAILED
            })
        ).toMatchSnapshot();
    });

    it('handles the reset state', () => {
        expect(
            reducer({} as any, {
                type: ActionNames.RESET_STATE
            })
        ).toMatchSnapshot();
    });
});
