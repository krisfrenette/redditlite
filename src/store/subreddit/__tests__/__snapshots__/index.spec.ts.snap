// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducer handles the fetch posts failed action 1`] = `
Object {
  "fetchingData": false,
}
`;

exports[`reducer handles the fetch posts request action 1`] = `
Object {
  "fetchingData": true,
}
`;

exports[`reducer handles the fetch posts success action 1`] = `
Object {
  "after": null,
  "before": null,
  "currentPaging": Object {
    "count": 0,
    "direction": null,
  },
  "fetchingData": false,
  "posts": Array [
    Object {
      "id": "mockId",
    },
  ],
}
`;

exports[`reducer handles the reset state 1`] = `
Object {
  "after": null,
  "before": null,
  "count": 0,
  "currentPaging": null,
  "fetchingData": true,
  "name": null,
  "posts": Array [],
}
`;

exports[`reducer sets a default state 1`] = `
Object {
  "after": null,
  "before": null,
  "count": 0,
  "currentPaging": null,
  "fetchingData": true,
  "name": null,
  "posts": Array [],
}
`;
